src/main/java/frc/lib/SwerveTrajectoryGenerator.java

```diff
@@ -0,0 +1,279 @@
+package frc.lib;
+
+import static frc.lib.SwerveTrajectory.MAXIMUM_LINEAR_ACCELERATION;
+import static frc.lib.SwerveTrajectory.MAXIMUM_LINEAR_VELOCITY;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+
+import frc.lib.SwerveTrajectory.Point;
+
+public class SwerveTrajectoryGenerator {
+    private static final double DISTANCE_BETWEEN_POINTS = 0.02;
+    private static final double SMOOTH_TOLERANCE        = 0.001;
+    private static final double SMOOTH_WEIGHT           = 0.99;
+    private static final double VELOCITY_PROPORTION     = 5;
+    private static final double VELOCITY_POWER          = 0.25;
+
+    public static SwerveTrajectory calculateTrajectory(Pose2d... waypoints) {
+        // Generate the coordinates of the path
+        Pose2d[] path = injectAndSmooth(waypoints);
+
+        double[] curvatures = new double[path.length];
+        double[] velocities = new double[path.length];
+        double[] distances = new double[path.length];
+        double[] predictedVelocities = new double[path.length];
+        calculateVelocities(path, curvatures, velocities, predictedVelocities,
+                distances);
+
+        double[] times = new double[path.length];
+        double[] accelerations = new double[path.length];
+        calculateAccelerations(predictedVelocities, distances, times,
+                accelerations);
+
+        Point[] points = constructPoints(path, curvatures, velocities,
+                distances, predictedVelocities, times, accelerations);
+        return new SwerveTrajectory(points);
+    }
+
+    private static Point[] constructPoints(Pose2d[] path, double[] curvatures,
+            double[] velocities, double[] distances,
+            double[] predictedVelocities, double[] times,
+            double[] accelerations) {
+        Point[] points = new Point[path.length];
+        for (int i = 0; i < path.length; i++) {
+            points[i] = new Point(path[i], curvatures[i], velocities[i],
+                    predictedVelocities[i], distances[i], times[i],
+                    accelerations[i]);
+        }
+        return points;
+    }
+
+    private static void calculateAccelerations(double[] predictedVelocities,
+            double[] distances, double[] times, double[] accelerations) {
+        double time = 0;
+        for (int i = 1; i < times.length; i++) {
+            double avgVelocity = 0.5
+                    * (predictedVelocities[i] + predictedVelocities[i - 1]);
+            double distance = distances[i] - distances[i - 1];
+            time += distance / avgVelocity;
+            times[i] = time;
+
+            double velocityChange = predictedVelocities[i]
+                    - predictedVelocities[i - 1];
+            double timeChange = times[i] - times[i - 1];
```

Comment:
If you were to name these variables `dv` and `dt` it will look a lot like calculus.